File: ios/RNAudienzzTargetingModule.m
#import "RNAudienzzTargetingModule.h"
#import <AudienzziOSSDK/AudienzziOSSDK-Swift.h>

@implementation RNAudienzzTargetingModule

RCT_EXPORT_MODULE();

- (dispatch_queue_t)methodQueue {
    return dispatch_get_main_queue();
}

RCT_EXPORT_METHOD(setOmidPartnerName:(NSString *)name) {
    [[AUTargeting shared] setOmidPartnerName:name];
}

RCT_EXPORT_METHOD(getOmidPartnerName:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *name = [[AUTargeting shared] omidPartnerName];
    resolve(name);
}

RCT_EXPORT_METHOD(setOmidPartnerVersion:(NSString *)version) {
    [[AUTargeting shared] setOmidPartnerVersion:version];
}

RCT_EXPORT_METHOD(getOmidPartnerVersion:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *version = [[AUTargeting shared] omidPartnerVersion];
    resolve(version);
}

RCT_EXPORT_METHOD(setPublisherName:(NSString *)name) {
    [[AUTargeting shared] setPublisherName:name];
}

RCT_EXPORT_METHOD(getPublisherName:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *name = [[AUTargeting shared] publisherName];
    resolve(name);
}

RCT_EXPORT_METHOD(setDomain:(NSString *)domain) {
    [[AUTargeting shared] setDomain:domain];
}

RCT_EXPORT_METHOD(getDomain:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *domain = [[AUTargeting shared] domain];
    resolve(domain);
}

RCT_EXPORT_METHOD(setStoreURL:(NSString *)url) {
    [[AUTargeting shared] setStoreURL:url];
}

RCT_EXPORT_METHOD(getStoreURL:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *url = [[AUTargeting shared] storeURL];
    resolve(url);
}

RCT_EXPORT_METHOD(setSourceapp:(NSString *)sourceapp) {
    [[AUTargeting shared] setSourceapp:sourceapp];
}

RCT_EXPORT_METHOD(getSourceapp:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *sourceapp = [[AUTargeting shared] sourceapp];
    resolve(sourceapp);
}

RCT_EXPORT_METHOD(setItunesID:(NSString *)itunesID) {
    [[AUTargeting shared] setItunesID:itunesID];
}

RCT_EXPORT_METHOD(getItunesID:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *itunesID = [[AUTargeting shared] itunesID];
    resolve(itunesID);
}

RCT_EXPORT_METHOD(setContentUrl:(NSString *)url) {
    [[AUTargeting shared] setContentUrl:url];
}

RCT_EXPORT_METHOD(getContentUrl:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *url = [[AUTargeting shared] contentUrl];
    resolve(url);
}

RCT_EXPORT_METHOD(setLatitude:(double)latitude longitude:(double)longitude) {
    [[AUTargeting shared] setLatitude:latitude longitude:longitude];
}

RCT_EXPORT_METHOD(setLocation:(double)latitude
                  longitude:(double)longitude
                  accuracy:(double)accuracy) {
    CLLocation *location = [[CLLocation alloc] initWithCoordinate:CLLocationCoordinate2DMake(latitude, longitude)
                                                         altitude:0
                                               horizontalAccuracy:accuracy
                                                 verticalAccuracy:0
                                                        timestamp:[NSDate date]];
    [[AUTargeting shared] setLocation:location];
}

RCT_EXPORT_METHOD(getLocation:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    CLLocation *location = [[AUTargeting shared] location];
    if (location) {
        NSDictionary *result = @{
            @"latitude": @(location.coordinate.latitude),
            @"longitude": @(location.coordinate.longitude),
            @"accuracy": @(location.horizontalAccuracy)
        };
        resolve(result);
    } else {
        resolve([NSNull null]);
    }
}

RCT_EXPORT_METHOD(setSubjectToCOPPA:(NSNumber *)isSubject) {
    [[AUTargeting shared] setCoppa:isSubject];
}

RCT_EXPORT_METHOD(getSubjectToCOPPA:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSNumber *coppa = [[AUTargeting shared] coppa];
    resolve(coppa);
}

RCT_EXPORT_METHOD(setSubjectToGDPR:(NSNumber *)isSubject) {
    [[AUTargeting shared] setSubjectToGDPR:isSubject];
}

RCT_EXPORT_METHOD(getSubjectToGDPR:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSNumber *gdpr = [[AUTargeting shared] getSubjectToGDPR];
    resolve(gdpr);
}

RCT_EXPORT_METHOD(setGdprConsentString:(NSString *)consent) {
    [[AUTargeting shared] setGdprConsentString:consent];
}

RCT_EXPORT_METHOD(getGdprConsentString:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *consent = [[AUTargeting shared] gdprConsentString];
    resolve(consent);
}

RCT_EXPORT_METHOD(setPurposeConsents:(NSString *)consents) {
    [[AUTargeting shared] setPurposeConsents:consents];
}

RCT_EXPORT_METHOD(getPurposeConsents:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *consents = [[AUTargeting shared] purposeConsents];
    resolve(consents);
}

RCT_EXPORT_METHOD(getDeviceAccessConsent:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSNumber *consent = [[AUTargeting shared] getDeviceAccessConsentObjc];
    resolve(consent);
}

RCT_EXPORT_METHOD(setExternalUserIds:(NSArray *)userIds) {
    [[AUTargeting shared] setEids:userIds];
}

RCT_EXPORT_METHOD(getExternalUserIds:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSArray *userIds = [[AUTargeting shared] getExternalUserIds];
    resolve(userIds ?: [NSNull null]);
}

RCT_EXPORT_METHOD(getUserKeywords:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSArray *keywords = [[AUTargeting shared] getUserKeywords];
    resolve(keywords);
}

RCT_EXPORT_METHOD(addUserKeyword:(NSString *)keyword) {
    [[AUTargeting shared] addUserKeyword:keyword];
}

RCT_EXPORT_METHOD(addUserKeywords:(NSArray *)keywords) {
    NSSet *keywordSet = [NSSet setWithArray:keywords];
    [[AUTargeting shared] addUserKeywords:keywordSet];
}

RCT_EXPORT_METHOD(removeUserKeyword:(NSString *)keyword) {
    [[AUTargeting shared] removeUserKeyword:keyword];
}

RCT_EXPORT_METHOD(clearUserKeywords) {
    [[AUTargeting shared] clearUserKeywords];
}

RCT_EXPORT_METHOD(getAppKeywords:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSArray *keywords = [[AUTargeting shared] getAppKeywords];
    resolve(keywords);
}

RCT_EXPORT_METHOD(addAppKeyword:(NSString *)keyword) {
    [[AUTargeting shared] addAppKeyword:keyword];
}

RCT_EXPORT_METHOD(addAppKeywords:(NSArray *)keywords) {
    NSSet *keywordSet = [NSSet setWithArray:keywords];
    [[AUTargeting shared] addAppKeywords:keywordSet];
}

RCT_EXPORT_METHOD(removeAppKeyword:(NSString *)keyword) {
    [[AUTargeting shared] removeAppKeyword:keyword];
}

RCT_EXPORT_METHOD(clearAppKeywords) {
    [[AUTargeting shared] clearAppKeywords];
}

RCT_EXPORT_METHOD(addUserData:(NSString *)key value:(NSString *)value) {
    [[AUTargeting shared] addUserDataWithKey:key value:value];
}

RCT_EXPORT_METHOD(updateUserData:(NSString *)key values:(NSArray *)values) {
    NSSet *valueSet = [NSSet setWithArray:values];
    [[AUTargeting shared] updateUserDataWithKey:key value:valueSet];
}

RCT_EXPORT_METHOD(getAppExtData:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSDictionary *extData = [[AUTargeting shared] getAppExtData];
    resolve(extData);
}

RCT_EXPORT_METHOD(addAppExtData:(NSString *)key value:(NSString *)value) {
    [[AUTargeting shared] addAppExtDataWithKey:key value:value];
}

RCT_EXPORT_METHOD(updateAppExtData:(NSString *)key values:(NSArray *)values) {
    NSSet *valueSet = [NSSet setWithArray:values];
    [[AUTargeting shared] updateAppExtDataWithKey:key value:valueSet];
}

RCT_EXPORT_METHOD(removeAppExtData:(NSString *)key) {
    [[AUTargeting shared] removeAppExtDataFor:key];
}

RCT_EXPORT_METHOD(clearAppExtData) {
    [[AUTargeting shared] clearAppExtData];
}

RCT_EXPORT_METHOD(getAccessControlList:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSArray *acl = [[AUTargeting shared] getAccessControlList];
    resolve(acl);
}

RCT_EXPORT_METHOD(addBidderToAccessControlList:(NSString *)bidderName) {
    [[AUTargeting shared] addBidderToAccessControlList:bidderName];
}

RCT_EXPORT_METHOD(removeBidderFromAccessControlList:(NSString *)bidderName) {
    [[AUTargeting shared] removeBidderFromAccessControlList:bidderName];
}

RCT_EXPORT_METHOD(clearAccessControlList) {
    [[AUTargeting shared] clearAccessControlList];
}

RCT_EXPORT_METHOD(addGlobalTargeting:(NSString *)key value:(NSString *)value) {
    [[AUTargeting shared] addGlobalTargetingWithKey:key value:value];
}

RCT_EXPORT_METHOD(addGlobalTargetingSet:(NSString *)key values:(NSArray *)values) {
    NSSet *valueSet = [NSSet setWithArray:values];
    [[AUTargeting shared] addGlobalTargetingWithKey:key values:valueSet];
}

RCT_EXPORT_METHOD(removeGlobalTargeting:(NSString *)key) {
    [[AUTargeting shared] removeGlobalTargetingWithKey:key];
}

RCT_EXPORT_METHOD(clearGlobalTargeting) {
    [[AUTargeting shared] clearGlobalTargeting];
}

RCT_EXPORT_METHOD(getGlobalOrtbConfig:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSString *config = [[AUTargeting shared] getGlobalOrtbConfig];
    resolve(config);
}

RCT_EXPORT_METHOD(setGlobalOrtbConfig:(NSString *)ortbConfig) {
    [[AUTargeting shared] setGlobalOrtbConfigWithOrtbConfig:ortbConfig];
}

RCT_EXPORT_METHOD(setUserExt:(NSDictionary *)userExt) {
    [[AUTargeting shared] setUserExt:userExt];
}

RCT_EXPORT_METHOD(getUserExt:(RCTPromiseResolveBlock)resolve
                  rejecter:(RCTPromiseRejectBlock)reject) {
    NSDictionary *userExt = [[AUTargeting shared] userExt];
    resolve(userExt ?: [NSNull null]);
}

@end

File: android/src/main/java/com/audienzzrn/RNAudienzzTargetingModule.kt
package com.audienzz

import android.util.Log
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactMethod
import org.audienzz.mobile.AudienzzTargetingParams
import android.util.Pair
import com.facebook.react.bridge.ReadableArray
import org.audienzz.mobile.AudienzzExternalUserId
import org.json.JSONObject

class RNAudienzzTargetingModule(reactContext: ReactApplicationContext) :
  ReactNativeModule(reactContext, SERVICE) {

  @ReactMethod
    fun setUserLatLng(latitude: Float, longitude: Float) {
      AudienzzTargetingParams.userLatLng = Pair(latitude, longitude)
    }

    @ReactMethod
    fun clearUserLatLng() {
      AudienzzTargetingParams.userLatLng = null
    }

    @ReactMethod
    fun getUserLatLng(promise: Promise) {
      val latLng = AudienzzTargetingParams.userLatLng
      if (latLng != null) {
        val result = Arguments.createMap()
        result.putDouble("latitude", latLng.first.toDouble())
        result.putDouble("longitude", latLng.second.toDouble())
        promise.resolve(result)
      } else {
        promise.resolve(null)
      }
    }

    @ReactMethod
    fun setPublisherName(name: String?) {
      AudienzzTargetingParams.publisherName = name
    }

    @ReactMethod
    fun getPublisherName(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.publisherName)
    }

    @ReactMethod
    fun setDomain(domain: String) {
      AudienzzTargetingParams.domain = domain
    }

    @ReactMethod
    fun getDomain(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.domain)
    }

    @ReactMethod
    fun setStoreUrl(url: String) {
      AudienzzTargetingParams.storeUrl = url
    }

    @ReactMethod
    fun getStoreUrl(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.storeUrl)
    }

    @ReactMethod
    fun setBundleName(bundleName: String?) {
      AudienzzTargetingParams.bundleName = bundleName
    }

    @ReactMethod
    fun getBundleName(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.bundleName)
    }

    @ReactMethod
    fun setOmidPartnerName(name: String?) {
      AudienzzTargetingParams.omidPartnerName = name
    }

    @ReactMethod
    fun getOmidPartnerName(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.omidPartnerName)
    }

    @ReactMethod
    fun setOmidPartnerVersion(version: String?) {
      AudienzzTargetingParams.omidPartnerVersion = version
    }

    @ReactMethod
    fun getOmidPartnerVersion(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.omidPartnerVersion)
    }

    @ReactMethod
    fun setSubjectToCOPPA(isSubject: Boolean?) {
      AudienzzTargetingParams.isSubjectToCOPPA = isSubject
    }

    @ReactMethod
    fun getSubjectToCOPPA(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.isSubjectToCOPPA)
    }

    @ReactMethod
    fun setSubjectToGDPR(isSubject: Boolean?) {
      AudienzzTargetingParams.isSubjectToGDPR = isSubject
    }

    @ReactMethod
    fun getSubjectToGDPR(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.isSubjectToGDPR)
    }

    @ReactMethod
    fun setGDPRConsentString(consent: String?) {
      AudienzzTargetingParams.gdprConsentString = consent
    }

    @ReactMethod
    fun getGDPRConsentString(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.gdprConsentString)
    }

    @ReactMethod
    fun setPurposeConsents(consents: String?) {
      AudienzzTargetingParams.purposeConsents = consents
    }

    @ReactMethod
    fun getPurposeConsents(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.purposeConsents)
    }

    @ReactMethod
    fun getDeviceAccessConsent(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.isDeviceAccessConsent)
    }

    @ReactMethod
    fun getUserKeywords(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.userKeywords)
    }

    @ReactMethod
    fun getKeywordSet(promise: Promise) {
      val keywords = AudienzzTargetingParams.keywordSet
      val array = Arguments.createArray()
      keywords.forEach { array.pushString(it) }
      promise.resolve(array)
    }

    @ReactMethod
    fun addUserKeyword(keyword: String) {
      AudienzzTargetingParams.addUserKeyword(keyword)
    }

    @ReactMethod
    fun addUserKeywords(keywords: ReadableArray) {
      val keywordSet = mutableSetOf<String>()
      for (i in 0 until keywords.size()) {
        keywords.getString(i)?.let { keywordSet.add(it) }
      }
      AudienzzTargetingParams.addUserKeywords(keywordSet)
    }

    @ReactMethod
    fun removeUserKeyword(keyword: String) {
      AudienzzTargetingParams.removeUserKeyword(keyword)
    }

    @ReactMethod
    fun clearUserKeywords() {
      AudienzzTargetingParams.clearUserKeywords()
    }


    @ReactMethod
    fun setExternalUserIds(userIds: ReadableArray?) {
      if (userIds == null) {
        AudienzzTargetingParams.setExternalUserIds(null)
        return
      }

      val externalUserIds = mutableListOf<AudienzzExternalUserId>()
      for (i in 0 until userIds.size()) {
        val userIdMap = userIds.getMap(i)
        val source = userIdMap?.getString("source") ?: continue
        val uniqueIdsArray = userIdMap.getArray("uniqueIds") ?: continue

        val uniqueIds = mutableListOf<AudienzzExternalUserId.AudienzzUniqueId>()
        for (j in 0 until uniqueIdsArray.size()) {
          val uniqueIdMap = uniqueIdsArray.getMap(j)
          val id = uniqueIdMap?.getString("id") ?: continue
          val atype = if (uniqueIdMap.hasKey("atype")) uniqueIdMap.getInt("atype") else continue

          uniqueIds.add(AudienzzExternalUserId.AudienzzUniqueId(id, atype))
        }

        externalUserIds.add(AudienzzExternalUserId(source, uniqueIds))
      }

      AudienzzTargetingParams.setExternalUserIds(externalUserIds)
    }

    @ReactMethod
    fun getExternalUserIds(promise: Promise) {
      val externalUserIds = AudienzzTargetingParams.getExternalUserIds()
      if (externalUserIds == null) {
        promise.resolve(null)
        return
      }

      val array = Arguments.createArray()
      externalUserIds.forEach { externalUserId ->
        val userIdMap = Arguments.createMap()
        userIdMap.putString("source", externalUserId.source)

        //TODO: add ability to get ids
//        val uniqueIdsArray = Arguments.createArray()
//        externalUserId.uniqueIds.forEach { uniqueId ->
//          val uniqueIdMap = Arguments.createMap()
//          uniqueIdMap.putString("id", uniqueId.id)
//          uniqueId.atype?.let { uniqueIdMap.putInt("atype", it) }
//          uniqueIdsArray.pushMap(uniqueIdMap)
//        }
//        userIdMap.putArray("uniqueIds", uniqueIdsArray)
        array.pushMap(userIdMap)
      }

      promise.resolve(array)
    }

    @ReactMethod
    fun getExtDataDictionary(promise: Promise) {
      val extData = AudienzzTargetingParams.extDataDictionary
      val map = Arguments.createMap()
      extData.forEach { (key, values) ->
        val array = Arguments.createArray()
        values.forEach { array.pushString(it) }
        map.putArray(key, array)
      }
      promise.resolve(map)
    }

    @ReactMethod
    fun addExtData(key: String, value: String) {
      AudienzzTargetingParams.addExtData(key, value)
    }

    @ReactMethod
    fun updateExtData(key: String, values: ReadableArray) {
      val valueSet = mutableSetOf<String>()
      for (i in 0 until values.size()) {
        values.getString(i)?.let { valueSet.add(it) }
      }
      AudienzzTargetingParams.updateExtData(key, valueSet)
    }

    @ReactMethod
    fun removeExtData(key: String) {
      AudienzzTargetingParams.removeExtData(key)
    }

    @ReactMethod
    fun clearExtData() {
      AudienzzTargetingParams.clearExtData()
    }

    @ReactMethod
    fun getAccessControlList(promise: Promise) {
      val acl = AudienzzTargetingParams.accessControlList
      val array = Arguments.createArray()
      acl.forEach { array.pushString(it) }
      promise.resolve(array)
    }

    @ReactMethod
    fun addBidderToAccessControlList(bidderName: String) {
      AudienzzTargetingParams.addBidderToAccessControlList(bidderName)
    }

    @ReactMethod
    fun removeBidderFromAccessControlList(bidderName: String) {
      AudienzzTargetingParams.removeBidderFromAccessControlList(bidderName)
    }

    @ReactMethod
    fun clearAccessControlList() {
      AudienzzTargetingParams.clearAccessControlList()
    }

    @ReactMethod
    fun addGlobalTargeting(key: String, value: String) {
      AudienzzTargetingParams.addGlobalTargeting(key, value)
    }

    @ReactMethod
    fun addGlobalTargetingSet(key: String, values: ReadableArray) {
      val valueSet = mutableSetOf<String>()
      for (i in 0 until values.size()) {
        values.getString(i)?.let { valueSet.add(it) }
      }
      AudienzzTargetingParams.addGlobalTargeting(key, valueSet)
    }

    @ReactMethod
    fun removeGlobalTargeting(key: String) {
      AudienzzTargetingParams.removeGlobalTargeting(key)
    }

    @ReactMethod
    fun clearGlobalTargeting() {
      AudienzzTargetingParams.clearGlobalTargeting()
    }

    @ReactMethod
    fun getGlobalOrtbConfig(promise: Promise) {
      promise.resolve(AudienzzTargetingParams.getGlobalOrtbConfig())
    }

    @ReactMethod
    fun setGlobalOrtbConfig(ortbConfig: String) {
      try {
        val jsonObject = JSONObject(ortbConfig)
        AudienzzTargetingParams.setGlobalOrtbConfig(jsonObject)
      } catch (e: Exception) {
        Log.e(TAG, "Invalid JSON for ORTB config: ${e.message}")
      }
    }

  companion object Companion {
    private const val SERVICE = "RNAudienzzTargetingModule"
    private const val TAG = "AudienzzTargetingModule"
  }
}
